File: CS/GridDemo.Wpf/ModuleResources/LargeDataSourceObjectBase.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;

namespace GridDemo {
    public abstract class LargeDataSourceObjectBase {
        protected class ValuesContainer<T> {
            Dictionary<int, T> modifiedValues;
            readonly Func<int, T> getDefaultValue;
            public ValuesContainer(Func<int, T> getDefaultValue) {
                this.getDefaultValue = getDefaultValue;
            }
            public T GetValue(int index) {
                if(modifiedValues == null)
                    return getDefaultValue(index);
                T value;
                if(modifiedValues.TryGetValue(index, out value))
                    return value;
                return getDefaultValue(index);
            }
            public void SetValue(int index, T value) {
                if(modifiedValues == null)
                    modifiedValues = new Dictionary<int, T>();
                modifiedValues[index] = value;
            }
        } 
        static readonly Priority[] Priorities = Enum.GetValues(typeof(Priority)).Cast<Priority>().ToArray();
        static LargeDataSourceObjectBase() {
            
        }

        static void GenerateCode() {
            var sb = new System.Text.StringBuilder();
            for(double i = 0; i < 1000d / 7d; i++) {
                var startIndex = i * 7 + 1;
                sb.AppendFormat(
@"
        [Display(Name = ""From ({0})"", Order = {0})]
        public string From{0} {{ get {{ return fromValues.GetValue({0}); }} set {{ fromValues.SetValue({0}, value); }} }}
        [Display(Name = ""To ({1})"", Order = {1})]
        public string To{1} {{ get {{ return toValues.GetValue({1}); }} set {{ toValues.SetValue({1}, value); }} }}
        [Display(Name = ""Sent ({2})"", Order = {2})]
        public DateTime Sent{2} {{ get {{ return sentValues.GetValue({2}); }} set {{ sentValues.SetValue({2}, value); }} }}
        [Display(Name = ""Has Attachment ({3})"", Order = {3})]
        public bool HasAttachment{3} {{ get {{ return hasAttachmentValues.GetValue({3}); }} set {{ hasAttachmentValues.SetValue({3}, value); }} }}
        [Display(Name = ""Size ({4})"", Order = {4})]
        public int Size{4} {{ get {{ return sizeValues.GetValue({4}); }} set {{ sizeValues.SetValue({4}, value); }} }}
        [Display(Name = ""Priority ({5})"", Order = {5}), GridEditor(TemplateKey = ""priorityColumn"")] 
        public Priority Priority{5} {{ get {{ return priorityValues.GetValue({5}); }} set {{ priorityValues.SetValue({5}, value); }} }}
        [Display(Name = ""Subject ({6})"", Order = {6}), GridEditor(TemplateKey = ""subjectEditor"")]
        public string Subject{6} {{ get {{ return subjectValues.GetValue({6}); }} set {{ subjectValues.SetValue({6}, value); }} }}
", startIndex, startIndex + 1, startIndex + 2, startIndex + 3, startIndex + 4, startIndex + 5, startIndex + 6);
            }
        }

        const int BaseColumnCount = 7;
        protected readonly ValuesContainer<string> fromValues;
        protected readonly ValuesContainer<string> toValues;
        protected readonly ValuesContainer<DateTime> sentValues;
        protected readonly ValuesContainer<bool> hasAttachmentValues;
        protected readonly ValuesContainer<int> sizeValues;
        protected readonly ValuesContainer<Priority> priorityValues;
        protected readonly ValuesContainer<string> subjectValues;
        public LargeDataSourceObjectBase(int id) {
            Id = id;
            fromValues = new ValuesContainer<string>(index => OutlookData.Users[GetPseudoRandomValue(Id, index, OutlookData.Users.Length)].Name);
            toValues = new ValuesContainer<string>(index => OutlookData.Users[GetPseudoRandomValue(Id, index + 5, OutlookData.Users.Length)].Name);
            sentValues = new ValuesContainer<DateTime>(index => DateTime.Today.AddDays(GetPseudoRandomValue(Id, index, 30)));
            hasAttachmentValues = new ValuesContainer<bool>(index => GetPseudoRandomValue(Id, index, 2) == 0 ? true : false);
            sizeValues = new ValuesContainer<int>(index => GetPseudoRandomValue(Id, index, 10000));
            priorityValues = new ValuesContainer<Priority>(index => Priorities[GetPseudoRandomValue(Id, index, Priorities.Length)]);
            subjectValues = new ValuesContainer<string>(index => OutlookDataGenerator.Subjects[GetPseudoRandomValue(Id, index, OutlookDataGenerator.Subjects.Length)]);
        }
        [Display(Name = "Id (0)", Order = 0)]
        public int Id { get; private set; }

        int GetPseudoRandomValue(int rowIndex, int columnIndex, int maxValue) {
            return (rowIndex + columnIndex) % maxValue;
        }
    }
}

File: VB/GridDemo.Wpf/ModuleResources/LargeDataSourceObjectBase.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel.DataAnnotations
Imports System.Linq

Namespace GridDemo
    Public MustInherit Class LargeDataSourceObjectBase
        Protected Class ValuesContainer(Of T)
            Private modifiedValues As Dictionary(Of Integer, T)
            Private ReadOnly getDefaultValue As Func(Of Integer, T)
            Public Sub New(ByVal getDefaultValue As Func(Of Integer, T))
                Me.getDefaultValue = getDefaultValue
            End Sub
            Public Function GetValue(ByVal index As Integer) As T
                If modifiedValues Is Nothing Then
                    Return getDefaultValue(index)
                End If
                Dim value As T = Nothing
                If modifiedValues.TryGetValue(index, value) Then
                    Return value
                End If
                Return getDefaultValue(index)
            End Function
            Public Sub SetValue(ByVal index As Integer, ByVal value As T)
                If modifiedValues Is Nothing Then
                    modifiedValues = New Dictionary(Of Integer, T)()
                End If
                modifiedValues(index) = value
            End Sub
        End Class
        Private Shared ReadOnly Priorities() As Priority = System.Enum.GetValues(GetType(Priority)).Cast(Of Priority)().ToArray()
        Shared Sub New()

        End Sub

        Private Shared Sub GenerateCode()
            Dim sb = New System.Text.StringBuilder()
            For i As Double = 0 To (1000R / 7R) - 1
                Dim startIndex = i * 7 + 1
                sb.AppendFormat("" & ControlChars.CrLf & _
"        [Display(Name = ""From ({0})"", Order = {0})]" & ControlChars.CrLf & _
"        public string From{0} {{ get {{ return fromValues.GetValue({0}); }} set {{ fromValues.SetValue({0}, value); }} }}" & ControlChars.CrLf & _
"        [Display(Name = ""To ({1})"", Order = {1})]" & ControlChars.CrLf & _
"        public string To{1} {{ get {{ return toValues.GetValue({1}); }} set {{ toValues.SetValue({1}, value); }} }}" & ControlChars.CrLf & _
"        [Display(Name = ""Sent ({2})"", Order = {2})]" & ControlChars.CrLf & _
"        public DateTime Sent{2} {{ get {{ return sentValues.GetValue({2}); }} set {{ sentValues.SetValue({2}, value); }} }}" & ControlChars.CrLf & _
"        [Display(Name = ""Has Attachment ({3})"", Order = {3})]" & ControlChars.CrLf & _
"        public bool HasAttachment{3} {{ get {{ return hasAttachmentValues.GetValue({3}); }} set {{ hasAttachmentValues.SetValue({3}, value); }} }}" & ControlChars.CrLf & _
"        [Display(Name = ""Size ({4})"", Order = {4})]" & ControlChars.CrLf & _
"        public int Size{4} {{ get {{ return sizeValues.GetValue({4}); }} set {{ sizeValues.SetValue({4}, value); }} }}" & ControlChars.CrLf & _
"        [Display(Name = ""Priority ({5})"", Order = {5}), GridEditor(TemplateKey = ""priorityColumn"")] " & ControlChars.CrLf & _
"        public Priority Priority{5} {{ get {{ return priorityValues.GetValue({5}); }} set {{ priorityValues.SetValue({5}, value); }} }}" & ControlChars.CrLf & _
"        [Display(Name = ""Subject ({6})"", Order = {6}), GridEditor(TemplateKey = ""subjectEditor"")]" & ControlChars.CrLf & _
"        public string Subject{6} {{ get {{ return subjectValues.GetValue({6}); }} set {{ subjectValues.SetValue({6}, value); }} }}" & ControlChars.CrLf & _
"", startIndex, startIndex + 1, startIndex + 2, startIndex + 3, startIndex + 4, startIndex + 5, startIndex + 6)
            Next i
        End Sub

        Private Const BaseColumnCount As Integer = 7
        Protected ReadOnly fromValues As ValuesContainer(Of String)
        Protected ReadOnly toValues As ValuesContainer(Of String)
        Protected ReadOnly sentValues As ValuesContainer(Of Date)
        Protected ReadOnly hasAttachmentValues As ValuesContainer(Of Boolean)
        Protected ReadOnly sizeValues As ValuesContainer(Of Integer)
        Protected ReadOnly priorityValues As ValuesContainer(Of Priority)
        Protected ReadOnly subjectValues As ValuesContainer(Of String)
        Public Sub New(ByVal id As Integer)
            Me.Id = id
            fromValues = New ValuesContainer(Of String)(Function(index) OutlookData.Users(GetPseudoRandomValue(Me.Id, index, OutlookData.Users.Length)).Name)
            toValues = New ValuesContainer(Of String)(Function(index) OutlookData.Users(GetPseudoRandomValue(Me.Id, index + 5, OutlookData.Users.Length)).Name)
            sentValues = New ValuesContainer(Of Date)(Function(index) Date.Today.AddDays(GetPseudoRandomValue(Me.Id, index, 30)))
            hasAttachmentValues = New ValuesContainer(Of Boolean)(Function(index)If(GetPseudoRandomValue(Me.Id, index, 2) = 0, True, False))
            sizeValues = New ValuesContainer(Of Integer)(Function(index) GetPseudoRandomValue(Me.Id, index, 10000))
            priorityValues = New ValuesContainer(Of Priority)(Function(index) Priorities(GetPseudoRandomValue(Me.Id, index, Priorities.Length)))
            subjectValues = New ValuesContainer(Of String)(Function(index) OutlookDataGenerator.Subjects(GetPseudoRandomValue(Me.Id, index, OutlookDataGenerator.Subjects.Length)))
        End Sub
        Private privateId As Integer
        <Display(Name := "Id (0)", Order := 0)>
        Public Property Id() As Integer
            Get
                Return privateId
            End Get
            Private Set(ByVal value As Integer)
                privateId = value
            End Set
        End Property

        Private Function GetPseudoRandomValue(ByVal rowIndex As Integer, ByVal columnIndex As Integer, ByVal maxValue As Integer) As Integer
            Return (rowIndex + columnIndex) Mod maxValue
        End Function
    End Class
End Namespace
